server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java

```diff
@@ -0,0 +1,205 @@
+    );
+
+    private final ClusterService clusterService;
+
+    public ShardLimitsHealthIndicatorService(ClusterService clusterService) {
+        this.clusterService = clusterService;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate(boolean verbose, int maxAffectedResourcesCount, HealthInfo healthInfo) {
+        var healthMetadata = HealthMetadata.getFromClusterState(clusterService.state());
+        if (healthMetadata == null || healthMetadata.getShardLimitsMetadata() == null) {
+            return unknownIndicator();
+        }
+
+        return mergeIndicators(
+            calculateForNormalNodes(healthMetadata.getShardLimitsMetadata()),
+            calculateForFrozenNodes(healthMetadata.getShardLimitsMetadata())
+        );
+    }
+
+    private HealthIndicatorResult mergeIndicators(StatusResult normalNodes, StatusResult frozenNodes) {
+        var finalStatus = HealthStatus.merge(Stream.of(normalNodes.status, frozenNodes.status));
+        var diagnoses = List.<Diagnosis>of();
+        var symptomBuilder = new StringBuilder();
+
+        if (finalStatus == HealthStatus.GREEN) {
+            symptomBuilder.append("The cluster has enough room to add new shards.");
+        }
+
+        // RED and YELLOW status indicates that the cluster might have issues. finalStatus has the worst between *normal and frozen* nodes,
+        // so we have to check each of the groups in order of provide the right message.
+        if (finalStatus.indicatesHealthProblem()) {
+            symptomBuilder.append("Cluster is close to reaching the maximum number of shards for ");
+            if (normalNodes.status == frozenNodes.status) {
+                symptomBuilder.append(NORMAL_GROUP).append(" and ").append(FROZEN_GROUP);
+                diagnoses = List.of(SHARD_LIMITS_REACHED_NORMAL_NODES, SHARD_LIMITS_REACHED_FROZEN_NODES);
+
+            } else if (normalNodes.status.indicatesHealthProblem()) {
+                symptomBuilder.append(NORMAL_GROUP);
+                diagnoses = List.of(SHARD_LIMITS_REACHED_NORMAL_NODES);
+
+            } else if (frozenNodes.status.indicatesHealthProblem()) {
+                symptomBuilder.append(FROZEN_GROUP);
+                diagnoses = List.of(SHARD_LIMITS_REACHED_FROZEN_NODES);
+            }
+
+            symptomBuilder.append(" nodes.");
+        }
+
+        var indicatorImpacts = switch (finalStatus) {
+            case RED -> RED_INDICATOR_IMPACTS;
+            case YELLOW -> YELLOW_INDICATOR_IMPACTS;
+            default -> List.<HealthIndicatorImpact>of();
+        };
+
+        return createIndicator(
+            finalStatus,
+            symptomBuilder.toString(),
+            buildDetails(normalNodes.result, frozenNodes.result),
+            indicatorImpacts,
+            diagnoses
+        );
+    }
+
+    private StatusResult calculateForNormalNodes(HealthMetadata.ShardLimits shardLimits) {
```

Comment:
this is my personal opinion, but I don't like mixing static and class methods much, in general. Having a mix of static and class methods becomes quickly messy in tests and code (having to pass a lot of parameters, potential side-effects, etc), from my experience. On the other hand, I think that **only** using side-effect-free static methods makes sense (like in functional programming), but since we already have a framework that ensures that we only have 1 self-contained instance of the object. 

Now, the methods are transitively tested when calling `calculate(...)`. I feel like exposing the methods just for testing it's a bit unneeded since we're not treating the class as a black box.

WDYT?

Reply:
++ I understand your reasoning.

We generally tend to favour static methods (functions) where we can as they're easier to reason about and they signal they don't use any global state (as you point out, there might be cases where a method would solely rely on the class global state and not only that but, business-wise they make sense to belong to the instance class). 

Testing these methods in isolation becomes much easier as well and the unit tests themselves, when not funnelled through the parent class, start to offer nice documentation around how the functions behave.
e.g. https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleService.java#L250

https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java#L976 

and 

https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SnapshotLifecycleServiceTests.java#L388